client/src/wayland_obj/manager.rs

```diff
@@ -0,0 +1,168 @@
+            } => wayland.new_global(registry, id, &interface, version),
+            _ => unimplemented!("GlobalEvent")
+        }
+    })
+}
+
+// TODO(ried): improve error handling: return something
+pub fn init_wayland() -> Result<(Display, EventQueue), ()> {
+    let connection = Display::connect_to_env();
+    if let Err(err) = connection {
+        use wayland_client::ConnectError::*;
+        match err {
+            NoWaylandLib => error!("Could not find Wayland library, is it installed and in PATH?"),
+            NoCompositorListening => {
+                error!("Could not connect to Wayland server. Is it running?");
+                error!(
+                    "WAYLAND_DISPLAY={}",
+                    std::env::var("WAYLAND_DISPLAY").unwrap_or_default()
+                );
+            },
+            InvalidName => error!("Invalid socket name provided in WAYLAND_SOCKET"),
+            XdgRuntimeDirNotSet => error!("XDG_RUNTIME_DIR must be set"),
+            InvalidFd => error!("Invalid socket provided in WAYLAND_SOCKET")
+        }
+        return Err(());
+    }
+
+    let (display, event_queue) = connection.unwrap();
+
+    WAYLAND.with(|wayland| {
+        let wayland = &mut wayland.borrow_mut();
+        let way_man = WaylandManager::new(crate::lua::OutputHandler);
+        (*wayland).replace(way_man);
+    });
+
+    Ok((display, event_queue))
+}
+
+impl WaylandManager {
+    fn new(output_handler: impl OutputEventHandler + 'static) -> Self {
+        let output_handler = Rc::new(output_handler);
+        WaylandManager {
+            compositor_manager: WlCompositorManager::default(),
+            output_manager: WlOutputManager::new(output_handler),
+            seat_manager: WlSeatManager::default(),
+            shell_manager: LayerShellManager::new(),
+            shm_manager: WlShmManager::new()
+        }
+    }
+
+    fn new_global(&mut self, registry: WlRegistry, id: u32, interface: &str, version: u32) {
+        match interface {
+            WlCompositor::NAME => Self::handle::<WlCompositor>(
+                registry,
+                WL_COMPOSITOR_VERSION,
+                version,
+                id,
+                &mut self.compositor_manager
+            ),
+            WlOutput::NAME => {
+                Self::handle::<WlOutput>(registry, WL_OUTPUT_VERSION, version, id, &mut self.output_manager)
+            },
+            WlSeat::NAME => {
+                Self::handle::<WlSeat>(registry, WL_SEAT_VERSION, version, id, &mut self.seat_manager)
+            },
+            WlShm::NAME => {
+                Self::handle::<WlShm>(registry, WL_SHM_VERSION, version, id, &mut self.shm_manager)
+            },
+            WlrLayerShell::NAME => Self::handle::<WlrLayerShell>(
+                registry,
+                WLR_LAYER_SHELL_VERSION,
+                version,
+                id,
+                &mut self.shell_manager
+            ),
+            _ => info!("unhandled global: {}", interface)
```

Comment:
I don't think this is worth logging. For debugging, it's already possible to see all globals using `weston-info`